pkg/core/mempool/notary_request_pool.go

```diff
@@ -0,0 +1,146 @@
+package mempool
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+	"sync"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/network/payload"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// NotaryRequestPool stores P2PNotaryRequest payloads.
+type NotaryRequestPool struct {
+	lock             sync.RWMutex
+	verifiedMap      map[util.Uint256]*payload.P2PNotaryRequest
+	verifiedPayloads []*payload.P2PNotaryRequest
+	fees             map[util.Uint160]utilityBalanceAndFees
+	capacity         int
+}
+
+// NewNotaryRequestPool returns a new NotaryRequestPool struct.
+func NewNotaryRequestPool(capacity int) *NotaryRequestPool {
+	return &NotaryRequestPool{
+		verifiedMap:      make(map[util.Uint256]*payload.P2PNotaryRequest),
+		verifiedPayloads: make([]*payload.P2PNotaryRequest, 0, capacity),
+		fees:             make(map[util.Uint160]utilityBalanceAndFees),
+		capacity:         capacity,
+	}
+}
+
+// HasPayload checks whether payload with the specified hash is in the pool.
+func (mp *NotaryRequestPool) HasPayload(h util.Uint256) bool {
+	mp.lock.RLock()
+	defer mp.lock.RUnlock()
+	return mp.containsKey(h)
+}
+
+func (mp *NotaryRequestPool) containsKey(hash util.Uint256) bool {
+	_, ok := mp.verifiedMap[hash]
+	return ok
+}
+
+// Add tries to add given payload to the NotaryRequestPool.
+func (mp *NotaryRequestPool) Add(r *payload.P2PNotaryRequest, d Depositor) error {
+	mp.lock.Lock()
+	defer mp.lock.Unlock()
+
+	if mp.containsKey(r.Hash()) {
+		return ErrDup
+	}
+
+	err := mp.checkPayloadConflicts(r, d)
+	if err != nil {
+		return fmt.Errorf("invalid P2PNotaryRequestPayload: %s", err.Error())
+	}
+
+	mp.verifiedMap[r.Hash()] = r
+
+	n := sort.Search(len(mp.verifiedPayloads), func(n int) bool {
+		// TODO: compare fallback transactions?
+		return r.MainTransaction.NetworkFee-mp.verifiedPayloads[n].MainTransaction.NetworkFee > 0
+	})
+
+	// We've reached our capacity already.
+	if len(mp.verifiedPayloads) == mp.capacity {
+		// Less prioritized than the least prioritized we already have, won't fit.
+		if n == len(mp.verifiedPayloads) {
```

Comment:
I'm thinking about making it simpler, like FIFO. It opens up some opportunity for mischief (`defaultP2PNotaryRequestPayloadPoolSize × NotaryServiceFeePerKey` GAS and you've filled the queue, but this GAS will burn in `MaxNotValidBeforeDelta` blocks freeing up the queue again, but it simplifies things also (allowing to drop sorted slice completely)). Though current scheme is susceptible to this kind of attack also, I can add any kind of fee to the main transaction that won't ever be completed.